server/rosetta/client_offline.go

```diff
@@ -113,109 +41,90 @@ func (c *Client) ConstructionPayload(_ context.Context, request *types.Construct
 		return nil, crgerrs.WrapError(crgerrs.ErrInvalidOperation, "expected at least one operation")
 	}
 
-	// convert rosetta operations to sdk msgs and fees (if present)
-	msgs, fee, err := opsToMsgsAndFees(c.ir, request.Operations)
+	tx, err := c.converter.ToSDK().UnsignedTx(request.Operations)
 	if err != nil {
 		return nil, crgerrs.WrapError(crgerrs.ErrInvalidOperation, err.Error())
 	}
 
-	metadata, err := getMetadataFromPayloadReq(request)
-	if err != nil {
-		return nil, crgerrs.WrapError(crgerrs.ErrBadArgument, err.Error())
+	metadata := new(ConstructionMetadata)
+	if err = metadata.FromMetadata(request.Metadata); err != nil {
+		return nil, err
 	}
 
-	txFactory := tx.Factory{}.WithAccountNumber(metadata.AccountNumber).WithChainID(metadata.ChainID).
-		WithGas(metadata.Gas).WithSequence(metadata.Sequence).WithMemo(metadata.Memo).WithFees(fee.String())
-
-	TxConfig := c.getTxConfig()
-	txFactory = txFactory.WithTxConfig(TxConfig)
-
-	txBldr, err := tx.BuildUnsignedTx(txFactory, msgs...)
+	txBytes, payloads, err := c.converter.ToRosetta().SigningComponents(tx, metadata, request.PublicKeys)
 	if err != nil {
 		return nil, err
 	}
 
-	// Sign_mode_legacy_amino is being used as default here, as sign_mode_direct
-	// needs the signer infos to be set before hand but rosetta doesn't have a way
-	// to do this yet. To be revisited in future versions of sdk and rosetta
-	if txFactory.SignMode() == signing.SignMode_SIGN_MODE_UNSPECIFIED {
-		txFactory = txFactory.WithSignMode(signing.SignMode_SIGN_MODE_LEGACY_AMINO_JSON)
-	}
-
-	signerData := authsigning.SignerData{
-		ChainID:       txFactory.ChainID(),
-		AccountNumber: txFactory.AccountNumber(),
-		Sequence:      txFactory.Sequence(),
-	}
+	return &types.ConstructionPayloadsResponse{
+		UnsignedTransaction: hex.EncodeToString(txBytes),
+		Payloads:            payloads,
+	}, nil
+}
 
-	signBytes, err := TxConfig.SignModeHandler().GetSignBytes(txFactory.SignMode(), signerData, txBldr.GetTx())
-	if err != nil {
-		return nil, err
+func (c *Client) PreprocessOperationsToOptions(_ context.Context, req *types.ConstructionPreprocessRequest) (response *types.ConstructionPreprocessResponse, err error) {
+	if len(req.Operations) == 0 {
+		return nil, crgerrs.WrapError(crgerrs.ErrBadArgument, "no operations")
 	}
 
-	txBytes, err := TxConfig.TxEncoder()(txBldr.GetTx())
+	// now we need to parse the operations to cosmos sdk messages
+	tx, err := c.converter.ToSDK().UnsignedTx(req.Operations)
```

Comment:
there's a lot of performance improvements that can be done (there's a double unmarshal on operations for example) but since I plan to do a follow up PR with extra tests immediately after this one is merged, I'd rather tackle the improvements in that, because this PR has got really big and its hard to review as it is. WDYT?